src/engine/Service.cpp

```diff
@@ -249,26 +226,27 @@ void Service::writeJsonResult(const std::vector<std::string>& vars,
 
   size_t rowIdx = 0;
   bool resultExists{false};
+  bool varsChecked{false};
   for (const nlohmann::json& part : body) {
     if (part.contains("head")) {
-      AD_CORRECTNESS_CHECK(part["head"].contains("vars") &&
-                           part["head"]["vars"].is_array());
-      verifyVariables(part);
+      AD_CORRECTNESS_CHECK(!varsChecked);
+      verifyVariables(part, body);
+      varsChecked = true;
     }
+    // The LazyJsonParser only yields parts containing the "bindings" array,
+    // therefore we can assume its existence here.
     AD_CORRECTNESS_CHECK(part.contains("results") &&
                          part["results"].contains("bindings") &&
                          part["results"]["bindings"].is_array());
-    auto bindings =
-        part["results"]["bindings"].get<std::vector<nlohmann::json>>();
-    writeBindings(bindings, rowIdx);
+    writeBindings(part["results"]["bindings"], rowIdx);
     resultExists = true;
   }
 
   // As the LazyJsonParser only passes parts of the result that match the
   // expected structure, no result implies an unexpected structure.
-  if (!resultExists) {
-    throw std::runtime_error(
-        "JSON result does not have the expected structure.");
+  if (!resultExists || !varsChecked) {
+    throwErrorWithContext("JSON result does not have the expected structure",
+                          body.details().first100_);
```

Comment:
You can make two distinct checks with distinct error messages
(`head section was missing`) or `results section was missing`.